Refactoring pipeline's resolving code

HiDAl opened the pull request:
The code to resolve pipelines has side effects that are not clear at first sight. 
The method `resolvePipelines` is renamed to `updateIndexRequestWithPipelines`, clearly stating its responsibilities. Also, it gets rid of the boolean returned value, which regarded whether the `indexRequest` has a pipeline defined and not whether the `indexRequest` was changed (what could have been deducted from the method signature).

This PR is in preparation for the bugfix of issue #83653 

Split of PR #92400.

Commit:
extract to its own method how to find the metadata

Commit:
rename some method making clear their objective

`resolvePipelines` method is actually mutating the passing object, what
wasn't clear from their name. Add a comment about and make it clear
within the method name

Commit:
avoid returning a potentially confusing value from `updateIndexRequestWithPipelines` method.

the returned value regards to whether the indexRequest has pipeline information, but, since it's a boolean, it could be confused at first sight with "I have updated the object".

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

Commit:
spotless fix

Commit:
add more unit tests

HiDAl commented:
Closing in favor of #93821 & #93863

joegallo reviewed (changes requested):
I'm not in favor of `findIndexMetadata` -- I don't think it's an especially good seam on which to break up this code.

At a high level, it looks to me like resolving pipelines is (at some level) fundamentally about **settings** -- if there's an index metadata, then grab the settings of that metadata, and use those settings to figure out the pipelines; if there's not an index metadata, run through any templates in order to see what the settings will be, then use those settings to figure out the pipelines.

It may not be possible to write a meaningful/useful 'findSettings' (due to the difference in how template settings resolution works as compared to index metadata settings), but perhaps something like a pair of methods along the lines of `updateIndexRequestWithPipelinesForExistingIndex` and `updateIndexRequestWithPipelinesViaTemplates` would be a better breakdown (both called by `updateIndexRequestWithPipelines`).

joegallo commented on server/src/test/java/org/elasticsearch/ingest/IngestServiceTests.java:
```diff
@@ -1877,15 +1908,13 @@ public void testResolveRequiredOrDefaultPipelineDefaultPipeline() {
         // index name matches with IDM:
         IndexRequest indexRequest = new IndexRequest("idx");
-        boolean result = IngestService.resolvePipelines(indexRequest, indexRequest, metadata);
-        assertThat(result, is(true));
```
It seems me like you're testing less than we used to -- all these lost `assertThat(result, is(true));` and `assertThat(result, is(false));` should instead be testing `indexRequestHasPipeline` now, right?

HiDAl replied:
That test is in line 1887 actually. but yeah, I'd add expand these tests with calls to that method

joegallo commented on server/src/main/java/org/elasticsearch/ingest/IngestService.java:
```diff
@@ -1223,4 +1211,37 @@ static class PipelineHolder {
+        return indexMetadata;
+    }
+
+    public static boolean indexRequestHasPipeline(IndexRequest indexRequest) {
```
I'm not sure, but perhaps this belongs on `IndexRequest` itself? Given that `IndexRequest` already has `isPipelineResolved`, maybe this makes sense there -- note that I'm not saying that it definitely does.

HiDAl replied:
Yes, I concur with you on this, but I'd do that in a different PR because I'd like to group several pipeline-related concepts in their own classes/records and clarify the domain. Let's discuss this offline later.

joegallo commented on server/src/main/java/org/elasticsearch/ingest/IngestService.java:
```diff
@@ -192,120 +192,108 @@ private static Map<String, Processor.Factory> processorFactories(List<IngestPlug
+     * @param indexRequest    The {@link org.elasticsearch.action.index.IndexRequest} object to update.
+     * @param clusterMetadata Cluster metadata from where the pipeline information could be derived.
+     */
+    public static void updateIndexRequestWithPipelines(
```
`resolvePipelines` was a good name, and I think it should come back.

The relationship between `resolvePipelines` and `isPipelineResolved` should get second-to-top billing in the docstring, since they're so tightly intertwined with each other.

Tell me about why you're not having it return a boolean anymore? I'm not saying I agree or disagree, but I'd like to understand.

HiDAl replied:
my logic is: this is a mutate-this-object-pls™️ method, so, the return value could be confused with a `Hey, I **did/didn't** change your object`. That was my reasoning the first time I read about this method in `TransportBulkAction`.

Commit:
rename method to `hasPipeline`

Commit:
introduces `Pipelines` record

Commit:
apply PR recommendations:

uses findPipeline* methods to calculate the pipelines according different strategies

Commit:
rename method to `resolvePipelinesAndUpdateIndexRequest`

Commit:
revert to check whether the updated request has a pipeline

Commit:
spotless

HiDAl commented:
great idea about the Setting stuff. That'll enable me to introduce the `Pipelines` object within this PR

Commit:
conciliate method names

HiDAl commented:
@elasticmachine run elasticsearch-ci/bwc

joegallo committed:
Use explicit types here

There's enough pipeline things getting thrown around that I'd actually
like the extra information.

joegallo committed:
Twiddle this logic

IMHO it's ever so slightly more clear that the request pipeline is
used if there is one OR the default pipeline is used if there isn't.

joegallo committed:
Emphasize that these values and blocks are distinct

The default and final pipeline either come entirely from one, or
entirely from the other. It's never the case that we mix and match.

joegallo committed:
Pipelines record can be private

joegallo committed:
Whitespace

joegallo committed:
Better use of settings defaults where possible

Rather than having defaulting at the Pipelines layer, just do
validation. Instead, rely on the defaulting at the settings layer,
where possible.

joegallo committed:
No need to rename this variable

The conventional name for a cluster metadata object of class Metadata
is just 'metadata'.

joegallo committed:
Fix this test name

joegallo committed:
Gah, typo

joegallo committed:
Merge branch 'main' into pull/93821

HiDAl commented on server/src/main/java/org/elasticsearch/ingest/IngestService.java:
```diff
@@ -1191,19 +1191,8 @@ private static Optional<Pipelines> resolvePipelinesFromMetadata(
-
-        return Optional.of(new Pipelines(defaultPipeline, finalPipeline));
+        final Settings settings = indexMetadata.getSettings();
+        return Optional.of(new Pipelines(IndexSettings.DEFAULT_PIPELINE.get(settings), IndexSettings.FINAL_PIPELINE.get(settings)));
```
oh nice! this method returns the default_value (`_none`) in case the settings does not exist. cool

joegallo commented on server/src/main/java/org/elasticsearch/ingest/IngestService.java:
```diff
@@ -1223,4 +1160,97 @@ static class PipelineHolder {
+        return Optional.of(new Pipelines(defaultPipeline, finalPipeline));
+    }
+
+    public static boolean hasPipeline(IndexRequest indexRequest) {
```
It's probably wrong to ever invoke this against an IndexRequest where `indexRequest.isPipelineResolved` is `false`, right? Given that, I'm inclined to add an assert for that. That'll probably require that you rewrite your tests (slightly).

Or, if it's not wrong (but I really do think it is!) then your tests need to cover null cases, too, but I really don't like that...

Along the same lines, if `indexRequest.isPipelineResolved` is asserted to be `true`, then it must never be the case that the pipeline fields are themselves null, which although the code is indeed null safe already might be worth also asserting (maybe?).

Given those asserts, it's probably logical to add a docstring.

Commit:
improve documentation of `resolvePipelinesAndUpdateIndexRequest` method

Commit:
assert potential inconsistent states in `hasPipelines` method

Commit:
reword comment

joegallo committed:
Merge branch 'main' into pull/93821

joegallo reviewed (approved):
LGTM
